synapse/lib/storm.lark

```diff
@@ -432,8 +432,121 @@ EXPRTIMES.2: "*"
-HEXNUMBER.1: /-?0x[0-9a-fA-F]+(?![^\s})*\/%+\-=><!~]|(?>(?<HEXNUMBERRECUR>\(((?>[^()"'`]+|(?&HEXNUMBERRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
+
+// Each of NUMBER, HEXNUMBER, and OCTNUMBER below have some very gnarly regular expressions. To make
+// them a little more palatable, they've been broken out and commented. All of these regex patterns
+// have a type specific section at the top for matching on their respective number patterns. The
+// negative lookahead is common to all three types. Basically all of this recursive nastiness is
+// following along to figure out if there is eventually a comma that turns it into an unquoted string
+// rather than a number. This is because storm allows for unquoted strings in its syntax. So, we need
+// to make sure we are only matching on number patterns that are not part of an unquoted string.
+// The atomic grouping stuff there makes sure that we are only matching entire levels of nesting at
+// a time, otherwise in something like this example:
+//
+//    if ($foo = 123 or $lib.str.concat('foo),b"ar', 'baz')) { }
+//
+// It will end up closing that level of parenthesis after "foo". This all means we need to have
+// these really funky regex patterns.
+
+NUMBER.1: /
+    -?                                          # optional sign
+    (                                           # start capture group
+        0+\d*(?=\.)                         |   # match zero, zero-leading numbers, and zero-leading numbers ending with . (e.g. "42.")
+        (?<!\d)[1-9]\d*                     |   # match non-zero-leading numbers 
+        0+(?![1-9\.])                           # match one or more zeroes
+    )                                           # end capture group
+    (\.\d*)?                                    # also capture decimal point and any following decimal digits
+
+    ##
+    ## The regex below to the end of the pattern is all the common stuff described above
+    ##
+    (?!                                         # start negative lookahead
+        [^\s})*\/%+\-=><!~]                 |   # exclude these characters from the negative lookahead
+        (?>                                     # atomic group
+            (?<NUMBERRECUR>                     # start of named capture group NUMBERRECUR
+                \((                             # literal open parens and start of capture group
+                    (?>                         # atomic group
+                        [^()"'`]+           |   # match anything except these characters
+                        (?&NUMBERRECUR)     |   # matches expression defined in capture group NUMBERRECUR
```

Comment:
I'd adjust this comment to something like "match another level of nested parentheses" for clarity because that is what the NUMBERRECUR capture group is doing